spec/serializers/user_serializer_spec.rb

```diff
@@ -73,6 +73,8 @@
 
   describe "private user attributes" do
     let(:result) do
+      # save user so that User.where/User.find can find zooniverse id
+      user.save
```

Comment:
user should have been created (`create(:user)`) by the factory and should be persisted already (have an id for lookup). this seems strange.

Reply:
This one was weird. We had a failing test https://github.com/zooniverse/panoptes/blob/master/spec/serializers/user_serializer_spec.rb#L87-L94 where `result[:zooniverse_id]` is nil. 

This is because https://github.com/zooniverse/panoptes/blob/e98eb1a80f247fbef465cd1e952b755eb6ef2071/spec/serializers/user_serializer_spec.rb#L76 we call `User.where(id: user.id)`. 

**Weird behavior is that when we either `user.reload`, `User.find(user.id)`, or `User.where(id:user.id)`, the result of `user.zooniverse_id` is `nil`. _BUT_ before the `user` is reloaded (or queried), `user` had a `zooniverse_id`  from the `create(:user)` & https://github.com/zooniverse/panoptes/blob/e98eb1a80f247fbef465cd1e952b755eb6ef2071/app/models/user.rb#L58.** 

My guess is that resulting user from the `after_create` is not saved in test.

Reply:
this is due to the `after_create` model callback (always a terrible idea to use these IMHO**)

A simple fix would be to set the zooniverse_id to what you expect on create, i.e. on line 6 setup the factory with the `zooniverse_id` already set to a random uniq value so it doesn't do anything on the callback
```
let(:user) { create(:user, migrated: migrated, sign_in_count: sign_in, zooniverse_id: "zoo-id-#{SecureRandom.random_number(100)}") }
```

** callbacks modify the model state without explicit action which you may not expect, id’ rather see this logic extracted out to a service object / operation style class that does this work for you…TBH this whole fix for ouroboros flow is terrible and should have been extracted to a service object from the start for use in the controller